### PR TITLE
arf-80297 load ogc_number and poa_codes into the RepresentatveUserLoader

### PR DESCRIPTION
**Pull request by gabezurita**
## Summary

- This work is behind a feature toggle (flipper): YES

## Related issue(s)
[ZH Issue](https://app.zenhub.com/workspaces/accredited-representative-facing-team-65453a97a9cc36069a2ad1d6/issues/gh/department-of-veterans-affairs/va.gov-team/80297)

Migration PR:
https://github.com/department-of-veterans-affairs/vets-api/pull/16494

## Testing done
- See specs

## What areas of the site does it impact?
- The Accredited Representative Portal
- 
## Acceptance criteria
 Ensure a representative user is assigned an OGC number that accurately retrieves their associated POA codes.

## Requested Feedback
How do we feel about this approach?

### CODE REVIEW

File: modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb
Comment (amprokop) [inline modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb:22]: FYI, @raudabaugh found that a small number of reps currently share email addresses in the OGC data set. We're thinking we should require reps to have unique email addresses to log into ARP, so we should code in an error case for this that will result in a message being shown to the user

**Review by amprokop — DISMISSED** [review]
LGTM, one comment

File: modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb
Comment (amprokop) [reply modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb:12]: Yep - is this the place in the code where we would see if any of the emails on the user's account are contained in the OGC list?

File: db/schema.rb
Comment (RachalCassity) [inline db/schema.rb:13]: Please pull from master. There will be a merge conflict

**Review by RachalCassity — CHANGES_REQUESTED** [review]

File: modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb
Comment (bramleyjl) [reply modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb:12]: The OGC list is represented in vets-api by `Veteran::Service::Representative` objects that are populated by the rake task, correct? If so then we have a [draft PR](https://github.com/department-of-veterans-affairs/vets-api/pull/16084) that I will work on refactoring after this change goes through to address list matching.

**Review by bramleyjl — COMMENTED** [review]
The connection looks good, I authenticated as a RepresentativeUser and was able to see the linked `VerifiedRepresentative` class and PoA code from the backing `AccreditedIndividual` class:

```ruby
[3] pry(#<AccreditedRepresentativePortal::RepresentativeUserLoader>)> @verified_representative
=> #<AccreditedRepresentativePortal::VerifiedRepresentative:0x00007502dd7d1298
 id: 1,
 ogc_registration_number: "12345",
 first_name: "John",
 last_name: "Doe",
 middle_initial: nil,
 email: "vets.gov.user+1@gmail.com",
 created_at: Thu, 25 Apr 2024 18:23:31.329422000 UTC +00:00,
 updated_at: Thu, 25 Apr 2024 19:11:35.667264000 UTC +00:00>
[4] pry(#<AccreditedRepresentativePortal::RepresentativeUserLoader>)> @verified_representative.poa_codes
=> ["987"]
```

File: db/migrate/20240418233828_create_accredited_representative_portal_verified_representatives.rb
Comment (bosawt) [inline db/migrate/20240418233828_create_accredited_representative_portal_verified_representatives.rb:7]: Is there any reason to have the first/last/middle names here? If you're going to link based on email then you prob don't even need these fields

File: modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb
Comment (gabezurita) [reply modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb:24]: Done! See commit:
a66de491aa

Description:
- Implement error handling for scenarios where a VerifiedRepresentative has a non-unique or non-existent email address in the associated AccreditedIndividual records.
- For the pilot, log errors without raising exceptions to ensure a smooth user experience.
- In the long term, consider raising errors and displaying appropriate messages to users when such inconsistencies are encountered.

Discussion:
- @raudabaugh discovered that a small number of representatives currently share email addresses in the OGC data set.
- To maintain data integrity and simplify the login process for ARP, it is recommended to require representatives to have unique email addresses.
- Allowing non-unique email addresses introduces complexity in two directions:
  1. Using user-entered OGC registration numbers for authentication, which are unique but public, essentially trusting users to provide accurate information.
  2. Using first and last names for authentication, which are not unique and can change (e.g., due to marriage), potentially leading to loss of access to ARP.
- For the pilot, the current approach of logging errors without raising exceptions is sufficient.
- In the long term, consider implementing a probabilistic approach (similar to MPI) for identity matching based on names, rather than a deterministic one, to avoid painful access scenarios for users.

CC: @raudabaugh @amprokop

File: modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb
Comment (amprokop) [inline modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb:52]: Good approach to start.

File: modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb
Comment (amprokop) [inline modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb:60]: Naming thought: Will `AccreditedIndividual` and `VerifiedRepresentative` have two distinct meanings in the long-term? What's the need for them to be two distinct classes?

Comment (amprokop) [reply modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb:60]: Maybe one of them is an ARP specific class and one of them is a row in the OGC data set? The names kind of confuse me right now in case you feel motivated to think of some future-proofed naming, not going to block on this though

File: modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb
Comment (amprokop) [inline modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb:12]: typo: alternate

File: modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb
Comment (amprokop) [inline modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb:13]: 👍

**Review by amprokop — APPROVED** [review]
Looks good overall, some Qs about naming/abstractions

**Review by raudabaugh — APPROVED** [review]
LGTM pending #16494 merge! Checked out locally and tests passed 👍

File: modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb
Comment (raudabaugh) [reply modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb:60]: Agree on the confusion here. I think renaming `AccreditedIndividual` to `Representative` would make its relationships to `RepresentativeUser` and `VerifiedRepresentative` more intuitive. Maybe some docstrings (like the one you have [here](https://github.com/department-of-veterans-affairs/vets-api/pull/16493/files#diff-3297a5571a7bdb7abf9835d3396fe6b6dd6c158c6f0b7f6316e3aa2d0c4d3ff7R15)) for `AccreditedIndividual` and `RepresentativeUser` describing those relationships would be helpful.

Comment (gabezurita) [reply modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb:60]: Good thoughts! To add my understanding of the usage of two models in our **`vets-api`**:

- **`AccreditedRepresentativePortal::VerifiedRepresentative`**: Intended exclusively for **`RepresentativeUsers`** manually verified by ARF as a provisional solution for the pilot. This is documented in the [[class-level docstring](https://github.com/department-of-veterans-affairs/vets-api/blob/e2b061b66e4733a2095d3f7b07b656196fd5d700/modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb#L4-L17)](https://github.com/department-of-veterans-affairs/vets-api/blob/e2b061b66e4733a2095d3f7b07b656196fd5d700/modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb#L4-L17).
- **`AccreditedIndividual`**: From my understanding, it represents every registered individual in the OGC dataset precisely as they appear there.

**Concerning** **`AccreditedIndividual`**:

- Given it currently serves as a direct mapping of the OGC data, introducing ARP-specific data fields might lead to confusion.
- As we know, a significant proportion (about 1/3) of the **`AccreditedIndividual`** records lack email addresses, complicating linking with **`RepresentativeUser`**. Any corrections to email addresses in the OGC dataset cannot be made on our side (whether `nil` or incorrect). Thus, we cannot rely on `AccreditedIndividual.find_by(email:)` to associate the two for the ARP pilot.
- Currently, more automated matching between **`RepresentativeUser`** and **`AccreditedIndividual`** would be solely done on the **`first`** and **`last name`,** which is insufficient.

**Possible Actions**:

- As Sam suggests, renaming the **`AccreditedIndividual`** model/table to reflect its role better and adding detailed docstrings seems like a good idea. I would love @holdenhinkle, @jvcAdHoc feedback on this (we can bring it up in the next ARM<>ARF sync).

File: modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb
Comment (gabezurita) [reply modules/accredited_representative_portal/app/services/accredited_representative_portal/representative_user_loader.rb]: Fixed! 0e7dd7c37a

File: db/schema.rb
Comment (RachalCassity) [inline db/schema.rb:131]: Please merge from master

**Review by RachalCassity — CHANGES_REQUESTED** [review]

File: db/schema.rb
Comment (RachalCassity) [reply db/schema.rb:131]: I will approve after merging from master

Comment (gabezurita) [reply db/schema.rb]: Thanks, @RachalCassity; just [merged main](https://github.com/department-of-veterans-affairs/vets-api/pull/16493/commits/b3def1cfe5a926e804c877420e572187e4fbf52d), which removed the migrations from this PR. Ready for review!

File: modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb
Comment (jvcAdHoc) [reply modules/accredited_representative_portal/app/models/accredited_representative_portal/verified_representative.rb:60]: I think this would be good to discuss at the next eng sync. For context, using the word "individual" was intentional to avoid overloading the word "representative" more than it already is.

I also agree about adding documentation to the model and plan to do so next time I'm working a ticket touching that code.

**Review by RachalCassity — APPROVED** [review]

**Pull request merged**
